enhancements/monitoring/alerting-consistency.md

```diff
@@ -0,0 +1,195 @@
+---
+title: alerting-consistency
+authors:
+  - "@michaelgugino"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2021-02-03
+last-updated: 2021-02-03
+status: implementable
+---
+
+# Alerting Consistency
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Alerts are a mess.  There's not really formal guidelines to follow when
+implementing alerts for a given component.  These are the guidelines.
+
+## Motivation
+
+Improve Signal to noise of alerts.  Align alert levels "Critical, Warning, Info"
+with operational expectations to allow easier triaging.  Ensure current and
+future alerts have documentation covering problem/investigation/solution steps.
+
+### Goals
+
+* Define clear criteria for designating an alert 'critical'
+* Define clear criteria for designating an alert 'warning'
+* Define minimum time thresholds before prior to triggering an alert
```

Comment:
That's unrealistic. Certain conditions can't be recovered from on their own and/or need instant action because in a high number of cases the component doesn't recover without interaction. There doesn't need to be an arbitrary wait time.
Generally arbitrary wait times are hard to define all over OpenShift

Reply:
I disagree.  This is about consistency.  The point of this document is to discuss what we think alerting and related elements should look like.  There's no reason to not define a standard set of criteria.  Anyone that thinks they have a special use case is probably wrong.

Reply:
this seems like it would be easy to audit.  For example, alerts installed via the CVO that do not set `for`:

```console
$ oc adm release extract --to manifests quay.io/openshift-release-dev/ocp-release:4.7.0-rc.0-x86_64
$ for x in $(grep -lr '^kind: PrometheusRule' manifests); do yaml2json < "${x}" | jq -r 'if (. | type) == "object" then [.] else . end | .[] | select(.kind == "PrometheusRule").spec.groups[].rules[] | select(.alert != null and .for == null) | .labels.severity + " " + .alert + " " + .expr'; done | sort | uniq

critical MCDRebootError mcd_reboot_err > 0
warning FailingOperator csv_abnormal{phase="Failed"}
warning ImageRegistryStorageReconfigured increase(image_registry_operator_storage_reconfigured_total[30m]) > 0
warning KubeletHealthState mcd_kubelet_state > 2
warning MCDPivotError mcd_pivot_err > 0
warning SystemMemoryExceedsReservation sum by (node) (container_memory_rss{id="/system.slice"}) > ((sum by (node) (kube_node_status_capacity{resource="memory"} - kube_node_status_allocatable{resource="memory"})) * 0.9)
```

How do we feel about those?

Reply:
"SystemMemoryExceedsReservation" is definitely one that needs a threshold.  I have some questions about the utility of this particular alert, but that's outside the scope of this discussion.

Reply:
`SystemMemoryExceedsReservation` is [here][1] from machine-config-operator#2033, if you want to follow up on that alert.

[1]: https://github.com/openshift/machine-config-operator/blame/master/install/0000_90_machine-config-operator_01_prometheus-rules.yaml#L53

Reply:
pretty much any alert that spuriously fires(in the sense that no action is needed, it's going to be fine once the upgrade finishes) during upgrade today is a candidate for "this alert needs to wait longer before firing" imho.

i'm also very interested to see examples where the time it takes an admin to see, process, and take action to resolve an alert doesn't significantly dominate the time between the condition happening and the alert firing.  That is to say:

if it takes an admin 30 minutes to fix something they were alerted to, then does it matter whether we alerted them 0 minutes or 5 minutes after it started happening?  Particularly if by waiting 5 minutes we may be able to avoid alerting them at all (because the situation does resolve itself, such as in the case of most alerts caused by node reboots).

Reply:
I tried to capture some timeline standards in the individual alerting sections.  I would say, for critical alerts, they can probably be added to a 0m, 5m or 15m bucket.

0m bucket:  etcd quorum is lost, we don't know how long this cluster might be alive so fire right away and hope for the best.
5m bucket: etcd quorum member lost (1 of 3).  If it's a small network blip, hopefully this is covered by etcd's internal timeout, and 5 minutes should hopefully be enough to recover.
15m bucket:  Pretty much all other critical alerts, like 1/3 healthy API servers.

Back to the upgrade bit, we shouldn't fire any alerts during upgrade, however, I don't think timing is the core issue, at least for the critical alerts.  The vast majority of alerts could be Warning or lower severity, and 60m timeout.  This achieves a number of things.  First, it gives time for the situation to self-correct, or for a helper like MachineHealthChecks to remediate a problem.  Second, it allows us to know that it's probably not going to self correct.  Third, it keeps your alert screen from having tons of alerts at once, especially if there are critical alerts firing, we don't need the extra distractions.

Many of our alerts today are 'raw' alerts, IMO.  Things like 9/10 daemonsets are running for MCD.  Does this need an alert?  It would be nice if we can aggregate some things into node specific issues.  EG, if sdn-daemon is down on a given node, we should have a 'NodeUnhealthy' alert, and aggregate some conditions onto that screen.  I'm not sure how feasible this is today.  At the very least, each component's operator could detect the specifics about it's operands and a given node.  For instance, if a given node is under maintenance, and the daemonset is missing, that's expected.  The operator is the layer we can utilize to understand this state.

Reply:
Isn't the wording on this a but doubled now?